### PR TITLE
Week 18 - Project Mongo API, Isabella Markskog

### PR DESCRIPTION
**Pull request by IsabellaM5**
**Link to my deployed project**
https://project-mongo-api-isabellam5.herokuapp.com/

**My reflections on how this week's project turned out**
I did all the blue level requirements and some red level.

I think this week was a great continuation on the project from the previous week. I feel this project turned out great! 😀 For this week I also had an ambition to create a frontend to my API displaying a documentation for my API. But I was running a bit short on time so this is something that I'll have to get back to later and implement.

**Things I'd like to have clarified or explained in more detail**
Nothing at the moment really 😀

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,39 +1,137 @@
...
+    }
+    if (booksToSend.length === 0) {
+      res.status(404).send(`Sorry, could not find any books!`);
+    }
+  
+    res.json(booksToSend);
+  } catch (err) {
+    res.status(404).send({ error: "Page not found" });
+  }
+});
```
Comment (sandra-wallen) [inline server.js:97]: Nice that you managed to put all queries inside same GET function, instead of one function for each 💪

File: server.js [hunk 2]
```diff
@@ -1,39 +1,137 @@
...
+    res.status(400).json({ error: 'Invalid book ID' });
+  }
+});
 
 // Start the server
 app.listen(port, () => {
-  // eslint-disable-next-line
-  console.log(`Server running on http://localhost:${port}`)
-})
+});
```
Comment (sandra-wallen) [inline server.js:137]: Great that you removed the console log 👏

**Review by sandra-wallen — APPROVED** [review]
Hi!

Great job on this project, you've implemented a lot of nice endpoints and queries!
As always, you have thought about coding defensive and user friendly 🤟. I really like that you put some documentation in your readme, it got me inspired to do so as well going forward!

Also, making sure that the connection to the database is up is really nice!

Well done! 🎉

File: README.md [hunk 3]
```diff
@@ -1,13 +1,54 @@
...
+
+### QUERIES
+
+Use endpoint ```GET /books```
+
+Filter by either author or title (optional). Case insensitive.
+
+* Example query for author: ```GET /books?author=adams```
+* Example query for title: ```GET /books?title=bill```
+* Example query for isbn: ```GET /books?isbn=439785960```
```
Comment (KarolCislowski) [inline README.md:54]: Nice documentation 😻

File: server.js [hunk 4]
```diff
@@ -1,39 +1,137 @@
...
+app.use(cors());
+app.use(express.json());
+
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next();
+  } else {
+    res.status(503).json({ error: 'Service unavailable' });
+  }
+});
```
Comment (KarolCislowski) [inline server.js:59]: Nice connection check-out 😻

File: server.js [hunk 5]
```diff
@@ -1,39 +1,137 @@
...
+    } 
+    if (title) {
+      booksToSend = await Book.find({ title: new RegExp(title, "i") });
+    }
+    if (isbn) {
+      booksToSend = await Book.find({ isbn: new RegExp(isbn, "i") });
+    }
+    if (booksToSend.length === 0) {
+      res.status(404).send(`Sorry, could not find any books!`);
+    }
```
Comment (KarolCislowski) [inline server.js:91]: great filtering 👍🏻

File: server.js [hunk 6]
```diff
@@ -1,39 +1,137 @@
...
+    res.json(booksToSend);
+  } catch (err) {
+    res.status(404).send({ error: "Page not found" });
+  }
+});
+
+// Books by top rating (4 or higher)
+app.get('/books/toprating', async (req, res) => {
+  try {
+    const booksByTopRating = await Book.find({ average_rating: { $gte: 4 } });
```
Comment (KarolCislowski) [inline server.js:102]: 👍🏻

File: server.js [hunk 7]
```diff
@@ -1,39 +1,137 @@
...
+
+  try {
+    const singleBook = await Book.findById(bookId);
+    if (singleBook) {
+      res.json(singleBook);
+    } else {
+      res.status(404).json({ error: 'Book not found' });
+    }
+  } catch (err) {
+    res.status(400).json({ error: 'Invalid book ID' });
```
Comment (KarolCislowski) [inline server.js:131]: nice choice of error status codes 🤓

**Review by KarolCislowski — APPROVED** [review]
Great API 🥇

**Pull request closed without merging**